### PR TITLE
WIP #6: add absolute() method on PathArc, still need tests

### PR DESCRIPTION
**Pull request by vitiral**
This is a first attempt at implementing #6.

### CODE REVIEW

**Comment by vitiral** [comment]
I conveted the whole library to using `absolute` for the default `PathAbs` instead of `canonicalize`. All tests passed except the ones I expected to fail and new unit tests are also looking good.

I'd say we are nearly there :smile:

**Comment by vitiral** [comment]
One thing that came out of this is that I don't actually _have_ to ensure that the path exists, only that it _could_ be a valid path given the cwd (i.e. you can't `..` more times than there are components in the cwd). Any path that starts with `/` (or windows prefix) will just call canonicalize on the first part of the path... which `/` _better_ exist :smile:

**Comment by vitiral** [comment]
bors try

**Comment by vitiral** [comment]
bors try

File: src/file.rs [hunk 1]
```diff
@@ -375,21 +375,26 @@ impl PathFile {
     ///
     /// let contents = "This is some contents";
     /// file.write_str(contents);
-    /// file.symlink(example_sym)?;
-    /// let file_sym = PathFile::new(example_sym)?;
+    /// let file_sym = file.symlink(example_sym)?;
     ///
     /// // They are canonicalized to the same file.
```
Comment (ExpHP) [inline src/file.rs]: Typo here? (Missing "not" or something?)

File: src/file.rs [hunk 2]
```diff
@@ -375,21 +375,26 @@ impl PathFile {
...
+    /// let file_can = file_sym.canonicalize()?;
+    /// assert_eq!(file, file_can);
     /// # Ok(()) } fn main() { try_main().unwrap() }
     /// ```
-    pub fn symlink<P: AsRef<Path>>(&self, dst: P) -> Result<()> {
+    pub fn symlink<P: AsRef<Path>>(&self, dst: P) -> Result<PathFile> {
         symlink_file(&self, &dst).map_err(|err| {
             Error::new(
                 err,
                 &format!("linking to {} from", dst.as_ref().display()),
```
Comment (ExpHP) [inline src/file.rs]: (while we're here, this message is confusing. Shouldn't it be `linking from {} to`?

**Comment by vitiral** [comment]
bors try

**Comment by vitiral** [comment]
@ExpHP I think since `PathAbs` now uses `PathArc::absolute()` as it's default `new` constructor we should be _mostly_ good.

I'm going to try and find all uses of canonicalize and make sure I'm not making any assumptions anywhere.

**Comment by vitiral** [comment]
bors try

**Comment by vitiral** [comment]
bors try

**Comment by vitiral** [comment]
bors try

**Comment by vitiral** [comment]
man, this is pure hell without a windows machine. I need to get a virtual machine working on this box.

**Comment by vitiral** [comment]
bors try

**Comment by vitiral** [comment]
bors try

**Pull request merged**